assets/js/base/hocs/with-reviews.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * External dependencies
+ */
+import { Component } from 'react';
+import { __, _n, sprintf } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import PropTypes from 'prop-types';
+
+/**
+ * Internal dependencies
+ */
+import { getReviews } from '../../blocks/reviews/utils';
+
+const withReviews = ( OriginalComponent ) => {
+	class WrappedComponent extends Component {
+		constructor() {
+			super( ...arguments );
+
+			this.state = {
+				error: null,
+				loading: false,
+				reviews: [],
+				totalReviews: 0,
+			};
+
+			this.setError = this.setError.bind( this );
+			this.delayedAppendReviews = this.props.delayFunction( this.appendReviews, 400 );
+		}
+
+		componentDidMount() {
+			this.replaceReviews();
+		}
+
+		componentDidUpdate( prevProps ) {
+			if ( prevProps.reviewsToDisplay < this.props.reviewsToDisplay ) {
+				// Since this attribute is controlled with a slider,
+				// it's better not to load the reviews immediately.
+				this.delayedAppendReviews();
+			} else if (
+				prevProps.categoryIds !== this.props.categoryIds ||
+				prevProps.orderby !== this.props.orderby ||
+				prevProps.order !== this.props.order ||
+				prevProps.productId !== this.props.productId
+			) {
+				this.replaceReviews();
+			}
+		}
+
+		getArgs() {
+			const { categoryIds, order, orderby, productId } = this.props;
+			const args = {
+				order,
+				orderby,
+			};
+
+			if ( categoryIds ) {
+				args.category_id = Array.isArray( categoryIds ) ? categoryIds.join( ',' ) : categoryIds;
+			}
+
+			if ( productId ) {
+				args.product_id = productId;
+			}
+
+			return args;
+		}
+
+		replaceReviews() {
+			const { announceUpdates, reviewsToDisplay } = this.props;
+			const args = {
+				...this.getArgs(),
+				per_page: reviewsToDisplay,
+			};
+
+			this.updateListOfReviews( args ).then( () => {
+				if ( announceUpdates ) {
+					speak( __( 'Reviews list updated.', 'woo-gutenberg-products-block' ) );
+				}
+			} );
+		}
+
+		appendReviews() {
+			const { announceUpdates, reviewsToDisplay } = this.props;
+			const { reviews } = this.state;
+
+			// Given that this function is delayed, props might have been updated since
+			// it was called so we need to check again if fetching new reviews is necessary.
+			if ( reviewsToDisplay <= reviews.length ) {
+				return;
+			}
+
+			const args = {
+				...this.getArgs(),
+				per_page: reviewsToDisplay - reviews.length,
+				offset: reviews.length,
+			};
```

Comment:
I wonder if we should just shift this to `updateListOfReviews`.  With the changes, it doesn't seem like `updateListOfReviews` needs to receive args as an argument anymore (nor reviews)?